### PR TITLE
Implement SyncBlock

### PR DESCRIPTION
**Pull request by drslebedev**
- This PR introduces a `SyncBlock` to synchronize data streams across multiple inputs that have the same sample rate. 
- It addresses key scenarios including initial time shifts, clock drift, and significant delays. 
- The `SyncBlock` synchronises input samples based on synchronization tags and publishes desynchronization tags with information about number of samples that were dropped.
- Add `bm_sync` for performance comparison

```
┌──────────────benchmark:───────────────┬──────┬──CPU branch misses───┬─ops/s─┬──mean──┬─<CPU-I>─┬───CPU cache misses───┬─stddev─┬─#N─┬─CTX-SW─┬─total time─┬─min─┬─median─┬─max─┐
│ src->gr::basic::SyncBlock<int>->sink  │ PASS │ 3.10M / 382M =  0.8% │  4.7G │ 843 ms │    930m │ 494M / 1.77G = 27.9% │        │    │     1  │     843 ms │     │        │     │
│ src->gr::basic::Selector<int>->sink   │ PASS │  394k / 363M =  0.1% │  4.7G │ 851 ms │    907m │ 409M / 1.86G = 22.0% │        │    │     2  │     851 ms │     │        │     │
│ src->copy->sink                       │ PASS │  403k / 125M =  0.3% │  4.8G │ 835 ms │    651m │ 337M / 1.94G = 17.3% │        │    │     2  │     835 ms │     │        │     │
└───────────────────────────────────────┴──────┴──────────────────────┴───────┴────────┴─────────┴──────────────────────┴────────┴────┴────────┴────────────┴─────┴────────┴─────┘
```

Related follow-up issue: #466

### CODE REVIEW

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            fmt::print("{}: configuration changed: n_ports {} -> {}\n", this->name, oldSettings.at("n_ports"), newSettings.at("n_ports"));
+            inputs.resize(n_ports);
+            outputs.resize(n_ports);
+            _nDroppedSamples.resize(n_ports, 0UZ);
+        }
+
+        if (newSettings.contains("max_history_size")) {
+            // should be less than actual buffer size (better < 80%)
+            // The logic has to be implemented when the new Port API is available
+        }
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:159]: As discussed in person: the block implementation should not assume nor use the Port's buffer API. 

I think it's OK to define a `max_history_size` that is used to define/communicate the required minimum buffer size to the graph and scheduler, and in turn be used implicitly by the block as the buffer having at least `max_history_size` slots.

However, the block should not depend on the size of the actually used Port-buffer. To note, we want to keep the buffer API slim and also connect blocks using other mechanisms than via buffers (<-> merge API).

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            if (!ins[i].isConsumeRequested()) {
+                ins[i].consume(0UZ);
+                ins[i].consumeTags(0UZ);
+                outs[i].publish(0UZ);
+            }
+        }
+        return gr::work::Status::OK;
+    }
+
+private:
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:232]: Consider removing the `private:` because having the functions below (most could be actually free-standing functions) public makes them easier to unit-test.

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    }
+
+    template<InputSpanLike TInput>
+    [[nodiscard]] constexpr std::vector<SyncData> synchronize(const std::span<TInput>& ins) {
+        const std::uint64_t syncTime = findSyncTime(ins);
+        if (syncTime == std::numeric_limits<std::uint64_t>::max()) {
+            return {};
+        }
+
+        std::vector<SyncData> syncData;
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:255]: At the risk that this might be a PMO: maybe we should consider making
```cpp
std::vector<SyncData> _syncData;
```
since this is called for every invocation of  `processBulk(..)? similarly for the `findSyncTime(...)` routine below. 

This may be likely a PMO but I am a bit concerned about two vectors being allocated everytime the block is invoked even if there are no synchronisation necessary. Maybe a basic benchmark (w/ and one w/o this block but a simple copy) that measures how much performance is lost might be useful. This way we could eliminate also future doubts and potential regressions.

In terms of benchmarks... guess we have plenty of `bm_,,,` to choose and copy-paste from. :-)

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            }
+        }
+        return in.size();
+    }
+
+    [[nodiscard]] bool isTimeDifferenceWithinTolerance(std::uint64_t t1, std::uint64_t t2) { return ((t1 > t2) ? t1 - t2 : t2 - t1) < tolerance; }
+
+    [[nodiscard]] constexpr bool isSyncTag(const gr::Tag& tag) const {
+        const std::string keyTriggerName = gr::tag::TRIGGER_NAME.shortKey();
+        const std::string keyTriggerTime = gr::tag::TRIGGER_TIME.shortKey();
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:339]: Why are these needed as a `std::string`? Doesn't the `at(..)` below work with string_views`

**Review by RalphSteinhagen — APPROVED** [review]
Thanks @drslebedev for this PR. This looks like a thorough, neat and concise implementation. I like it.

I think this could be already merged as-is but could benefit from some (optional) improvements outlined in the comments above.

Notably:
 1. the time-out and zero-padding isn't implemented, which is OK for now but should be marked by a `TODO` and followed up with a dedicated issue (we should discuss and meet for discussing the details since this is a wider issue/feature requirements)
  2. adding a benchmark to assess the cost of performing the synchronisation on streams that are already synchronised. This would help a lot when further refining the implementation and/or to detect potential performance regressions.
  
In any case: good job and well done! :+1:

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp (hunk 1 above)
Comment (drslebedev) [reply blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:159]: The idea here is to check if max_history_size is actually < buffer size, if not throw an gr::exception.

File: blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp (hunk 4 above)
Comment (drslebedev) [reply blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp:339]: `std::map::at` will not work with `std::sting_view`
